tool/tctl/common/device/resource.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Resource is a wrapper around devicepb.Device that implements types.Resource.
+type Resource struct {
+	// ResourceHeader is embedded to implement types.Resource
+	types.ResourceHeader
+	// Spec is the device specification
+	Spec ResourceSpec `json:"spec"`
+}
+
+// ResourceSpec is the device resource specification.
+type ResourceSpec struct {
+	// OsType is the device's operating system type.
+	OsType string `json:"os_type"`
+	// AssetTag is the device's asset tag.
+	AssetTag string `json:"asset_tag"`
+	// CreateTime is the time the device was created.
+	CreateTime time.Time `json:"create_time,omitempty"`
+	// UpdateTime is the time the device was last updated.
+	UpdateTime time.Time `json:"update_time,omitempty"`
+	// EnrollToken is the device's enrollment token.
+	EnrollToken *devicepb.DeviceEnrollToken `json:"enroll_token,omitempty"`
+	// EnrollStatus is the device's enrollment status.
+	EnrollStatus string `json:"enroll_status"`
+	// Credential is the device's credential.
+	Credential *devicepb.DeviceCredential `json:"credential,omitempty"`
+	// CollectedData is the device's collected data.
+	CollectedData []*devicepb.DeviceCollectedData `json:"collected_data,omitempty"`
+}
+
+// checkAndSetDefaults sanity checks Resource fields to catch simple errors, and
+// sets default values for all fields with defaults.
+func (r *Resource) checkAndSetDefaults() error {
+	if err := r.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if r.Kind == "" {
+		r.Kind = types.KindDevice
+		// Sanity check.
+	} else if r.Kind != types.KindDevice {
+		return trace.BadParameter("unexpected resource kind %q, must be %q", r.Kind, types.KindDevice)
+	}
+
+	if _, ok := devicepb.OSType_value[r.Spec.OsType]; !ok {
+		return trace.BadParameter("invalid OS type: %q", r.Spec.OsType)
+	}
+
+	if r.Spec.AssetTag == "" {
```

Comment:
Leave this to the backend?

If we do, we could allow empty OS type too.

Reply:
I believe we usually perform clientside validation to get better clientside UX due to error wrapping and for better latency. This is at least the norm in all our other resource interactions with CLI commands. Should we do away with this? I'm not sure but it'd seem wierd to stray here, that said this is a new way of working with resources purely for display, too.

Reply:
A minimal round of validations, like we have here, seems OK. I wouldn't do more, but it's fine by me if you'd rather keep it.